test/functional/client_side_encryption_corpus_tests.js

```diff
@@ -0,0 +1,381 @@
+          //    - ``method`` is either ``auto``, for automatic encryption or ``explicit`` for  explicit encryption
+          //    - ``identifier`` is either ``id`` or ``altname`` for the key identifier
+          //    - ``allowed`` is a boolean indicating whether the encryption for the given parameters is permitted.
+          //    - ``value`` is the value to be tested.
+          //
+          //    Create a new BSON document, named ``corpus_copied``.
+          //
+          //    Iterate over each field of ``corpus``.
+          //    - If the field name is ``_id``, ``altname_aws`` and ``altname_local``, copy the field to ``corpus_copied``.
+          //    - If ``method`` is ``auto``, copy the field to ``corpus_copied``.
+          //    - If ``method`` is ``explicit``, use ``client_encryption`` to explicitly encrypt the value.
+          //      - Encrypt with the algorithm described by ``algo``.
+          //      - If ``identifier`` is ``id``
+          //        - If ``kms`` is ``local`` set the key_id to the UUID with base64 value ``LOCALAAAAAAAAAAAAAAAAA==``.
+          //        - If ``kms`` is ``aws`` set the key_id to the UUID with base64 value ``AWSAAAAAAAAAAAAAAAAAAA==``.
+          //      - If ``identifier`` is ``altname``
+          //        - If ``kms`` is ``local`` set the key_alt_name to "local".
+          //        - If ``kms`` is ``aws`` set the key_alt_name to "aws".
+          //      If ``allowed`` is true, copy the field and encrypted value to ``corpus_copied``.
+          //      If ``allowed`` is false. verify that an exception is thrown. Copy the unencrypted value to to ``corpus_copied``.
+          return forEachP(Object.keys(corpus), key => {
+            const field = corpus[key];
+            if (copyOverValues.has(key)) {
+              corpusCopied[key] = field;
+              return;
+            }
+            if (field.method === 'auto') {
+              corpusCopied[key] = Object.assign({}, field);
+              return;
+            }
+            if (field.method === 'explicit') {
+              const encryptOptions = {
+                algorithm: algorithmMap.get(field.algo)
+              };
+              if (field.identifier === 'id') {
+                encryptOptions.keyId = identifierMap.get(field.kms);
+              } else if (field.identifier === 'altname') {
+                encryptOptions.keyAltName = keyAltNameMap.get(field.kms);
+              } else {
+                throw new Error('wtf how did u get here?');
+              }
+
+              return Promise.resolve()
+                .then(() => clientEncryption.encrypt(field.value, encryptOptions))
+                .then(
+                  encryptedValue => {
+                    if (field.allowed === true) {
+                      corpusCopied[key] = Object.assign({}, field, { value: encryptedValue });
+                    } else {
+                      throw new Error('This should have errored');
```

Comment:
nit: can we add a better error message here? "Expected encryption to fail for: value, blah blah"